tests/e2e/functional_test.go

```diff
@@ -30,12 +31,162 @@ type FunctionalSuite struct {
+	w.PushToGitRepo("basic-resources/modified", []string{"multiple-resources.yaml"}, false).Wait(30 * time.Second)
+	w.Expect().ResourcesExist("apps/v1", "deployments", []string{"multi-deploy"})
+	w.Expect().ResourcesExist("v1", "configmaps", []string{"multi-config"})
+	w.Expect().CheckCommitStatus()
+
+	w.PushToGitRepo("basic-resources/modified", []string{"multiple-resources.yaml"}, true).Wait(30 * time.Second)
+	w.Expect().ResourcesDontExist("apps/v1", "deployments", []string{"multi-deploy"})
+	w.Expect().ResourcesDontExist("v1", "configmaps", []string{"multi-config"})
+	w.Expect().CheckCommitStatus()
+
+}
+
+// GitSync self healing occurs when resource does not match manifest in repo
+func (s *FunctionalSuite) TestSelfHealing() {
+
+	w := s.Given().GitSync("@testdata/gitsync.yaml").InitializeGitRepo("basic-resources/initial-commit").
+		When().
+		CreateGitSyncAndWait()
+	defer w.DeleteGitSyncAndWait()
+
+	// verify that test deployment is created with {replicas: 3} in spec
+	w.Expect().ResourcesExist("apps/v1", "deployments", []string{"test-deploy"})
+	w.Expect().CheckCommitStatus()
+
+	w.Expect().VerifyResourceState("apps/v1", "deployments", "test-deploy", "spec", "replicas", 3)
+
+	// apply patch to resource
+	w.ModifyResource("apps/v1", "deployments", "test-deploy", `{"spec":{"replicas":4}}`).Wait(30 * time.Second)
+
+	// verify that resource has been "healed" by resetting replica count to 3
+	w.Expect().VerifyResourceState("apps/v1", "deployments", "test-deploy", "spec", "replicas", 3)
+
+}
+
+// test behavior when changing the repoUrl of a GitSync
+func (s *FunctionalSuite) TestChangeRepoUrl() {
+
+	w := s.Given().GitSync("@testdata/gitsync.yaml").InitializeGitRepo("basic-resources/initial-commit").
+		When().
+		CreateGitSyncAndWait()
+	defer w.DeleteGitSyncAndWait()
+
+	// first repo's manifests should be applied
+	w.Expect().ResourcesExist("apps/v1", "deployments", []string{"test-deploy"})
+	w.Expect().ResourcesExist("v1", "configmaps", []string{"test-config"})
+	w.Expect().ResourcesExist("v1", "secrets", []string{"test-secret"})
+	w.Expect().CheckCommitStatus()
+
+	// changing repoUrl to point to repo2.git
+	w = s.Given().GitSync("@testdata/revised-gitsync.yaml").InitializeGitRepo("basic-resources/second-repo").
+		When().
+		UpdateGitSyncAndWait()
+
+	// verify that resources in new repository are created
+	w.Expect().ResourcesExist("apps/v1", "deployments", []string{"repo-two-deploy"})
+	w.Expect().ResourcesExist("v1", "configmaps", []string{"repo-two-config"})
+	w.Expect().ResourcesExist("v1", "secrets", []string{"repo-two-secret"})
+	w.Expect().CheckCommitStatus()
+
+	// old resources are deleted as their manifests do not exist in new repository
+	w.Expect().ResourcesDontExist("apps/v1", "deployments", []string{"test-deploy"})
+	w.Expect().ResourcesDontExist("v1", "configmaps", []string{"test-config"})
+	w.Expect().ResourcesDontExist("v1", "secrets", []string{"test-secret"})
+
+}
+
+func (s *FunctionalSuite) TestKustomize() {
```

Comment:
@juliev0 Technical question for kustomize, is the GitSync watching for the resource.yaml file to be deleted or the resource being deleted from the kustomization.yaml files? (ie. kustomize file at the base dir and kustomize file where the resource files are)

Reply:
Let's talk about what _should_ be rather than what _is_ since that's what matters. Basically, Numaplane should run kustomize on the kustomization.yaml file, so whatever resources are named in there are what matters. Does it answer your question or no?

But one clarification - if the kustomization.yaml file references another repository and something changes in that repository, we won't be notified of the change or anything. So, we won't react to it. That was a design decision made I guess. But if something in our own repo changes, then we'll always run kustomize on whatever is latest and greatest, so that's when we'd notice a change made within that repo.

Reply:
That does answer my question then, I need to look into why the deployment for example isn't being deleted when it's removed from the kustomization file in that case.

Will the GitSync just run the kustomize file again when it sees any changes to the manifests on the repo?

Reply:
Good, maybe you've found another bug. Want to explore the code a little from here?: https://github.com/numaproj-labs/numaplane/blob/main/internal/sync/syncer.go#L232 You should find some code related to kustomize manifests in there

Reply:
Will do! If this is in fact a bug, should we look to get that fixed first before pushing this or getting this in sooner? The test could be commented out for now and be included in the followup PRs that I had discussed some with Hao.

Reply:
I'm okay either way. Just make sure a bug gets logged first

Reply:
I'm not positive just yet, I think this may be something Chandan is fixing here: https://github.com/numaproj-labs/numaplane/pull/216

Going to ask if he can test his PR with kustomize repo.

Reply:
@dpadhiar why do you need to do L184? Are you trying to remove `deployment.yaml`?

@chandankumar4 I believed you test if kustomize file changed the resources should be added/deleted/updated accordingly? And we are checking each time with kustomize build the end yaml file right? So even there is a remote referred repository we can detect the change, right?

Reply:
@xdevxy L184 is leftover from me testing if the resource wasn't being deleted because the manifest was still on the repo. That can be removed once @chandankumar4 change is merged.